File: algorithms/2_medium/M/02207_maximize_number_of_subsequences_in_a_string/solution.c
long long maximumSubsequenceCount(const char *restrict text, const char *pattern) {
    const char p1 = pattern[0];
    const char p2 = pattern[1];
    int64_t p1_cnt = 0;
    int64_t p2_cnt = 0;
    int64_t res = 0;

    for (size_t i = 0; text[i] != '\0'; i += 1) {
        if (text[i] == p2) {
            res += p1_cnt;
            p2_cnt += 1;
        }

        if (text[i] == p1) { p1_cnt += 1; }
    }

    return res + MAX(p1_cnt, p2_cnt);
}

File: algorithms/2_medium/M/02207_maximize_number_of_subsequences_in_a_string/solution.rs
impl Solution {
    pub fn maximum_subsequence_count(text: String, pattern: String) -> i64 {
        assert!(pattern.len() == 2);
        let (p1, p2) = (pattern.as_bytes()[0], pattern.as_bytes()[1]);
        let (mut p1_cnt, mut p2_cnt): (i64, i64) = (0, 0);
        let mut res: i64 = 0;

        for &byte in text.as_bytes() {
            if byte == p2 {
                res += p1_cnt;
                p2_cnt += 1;
            }

            if byte == p1 { p1_cnt += 1; }
        }

        res + std::cmp::max(p1_cnt, p2_cnt)
    }
}
